service/history/shard/context_impl.go

```diff
@@ -726,15 +731,20 @@ func (s *ContextImpl) addTasksLocked(
 	}
 
 	request.RangeID = s.getRangeIDLocked()
-	err := s.executionManager.AddTasks(request)
-	if err = s.handleErrorAndUpdateMaxReadLevelLocked(err, transferMaxReadLevel); err != nil {
-		return err
+
+	// Drop lock while calling persistence to allow for more concurrency
+	s.wUnlock()
+	err = s.executionManager.AddTasks(request)
+	s.wLock()
+	if err == nil {
+		// Do this inside the lock because we don't want the transfer queue processor to wake up
```

Comment:
`handleErrorAndUpdateMaxReadLevelLocked` (sorry about the name...) returns nil iff its argument is nil, so that can't happen. Specifically, it always returns the err passed in. I suppose that means there's no reason it should return a value at all, but I didn't change it because I figured we might want to do some error mapping some day. I can get rid of the return value if you think it's an improvement.

Reply:
I'd vote for "simpler is better" ... add the return value only when you actually need it.